instance/models/utils.py

```diff
@@ -608,6 +616,99 @@ def put(self, key, value, **kwargs):
 
         return self._client.kv.put(consul_key, value, **kwargs)
 
+    def create_or_update_dict(self, value, num_retries=3, **kwargs):
+        """
+        Create or update dictionary.
+        This method handles backwards compatibility when writing ocim configuration to consul's KV store.
+        The legacy to store configuration values is to use a prefix and store each key/value pair as a different record.
+        As of June/2019 we store configuration values using the instance's prefix as key and the value as a JSON object.
+        By using only one key we avoid the need to use transactions and to have extra logic to read/write the
+        configuration. We still need this method to make sure we increment the version number when the configuration
+        value is updated.
+
+        .. note::
+            Once all instances' configuration are updated we can retire
+            some of the code in this class.
+
+        .. note::
+            This method also works as a PATCH if only a partial payload is
+            given. To delete a key use :meth:`delete_dict_key`
+
+        :param (dict) value: value to write
+        :param (int) num_retries: Retries for self.txn_put
+        :param kwargs: consul.kv.put specific options.
+        """
+        get_data = None
+        try:
+            data_index, get_data = self._client.kv.get(self.prefix, **kwargs)
+        except consul.base.NotFound:
+            pass
+        try:
+            # This is only needed for backwards compatibility.
+            data_index, txn_data = self._client.kv.get(self.prefix, recurse=True, **kwargs)
+        except consul.base.NotFound:
+            pass
+
+        if not get_data and txn_data:
+            return self.txn_put(value, num_retries)
+
+        if get_data:
+            stored = json.loads(get_data['Value'].decode('utf-8'))
+            updates = {
+                k: value[k] for k in value
+                if (k in stored and
+                    value[k] != stored[k]) or
+                k not in stored
+            }
+            payload = {
+                k: stored[k] for k in stored
+                if k not in value and
+                k != 'version'
+            }
+            updated = bool(updates)
+            payload['version'] = stored['version']
+            payload.update(value)
+        else:
+            payload = value
```

Comment:
Shouldn't `txn_data` be updated here as well in case of a PATCH is applied to a prefix that hasn't been updated?

Reply:
`txn_data` will only be available if (and only if) the instance's metadata is saved in the old style (e.g. `ocim/instances/<id>/slug`, `ocmi/instances/<id>/domains`, etc...).
If that's the case then we don't create the new style of metadata and we just use `self.txn_update` to use the old code and update the metadata accordingly.
If there's no `txn_data` and there's no `get_data` then this is a new instance and the metadata is created in the new style. Which means just passing the dictionary.
In case this is a PATCH to a metadata object in the new style, that would mean that there will be some value in `get_data` and no value in `txn_data` and we only update the updated keys on the dictionary.

Reply:
> If that's the case then we don't create the new style of metadata and we just use self.txn_update to use the old code and update the metadata accordingly.

But when we have the old style metadata, this will have get both txn_data and get_data, right? Won't it be updated here?

Reply:
@viadanna If there's an old-style configuration in the metadata then only `txn_data` will get populated.
That's because `consul.kv.get` will match the key given exactly and in the old-style the instance's prefix is not stored in the kv store, e.g. `ocim/instsances/1/key1`, `ocim/instsances/1/key2`, etc...
So, if you have this in Ocim:
```json
[{
    "Key": "ocim/instances/123/key1",
    "Value": "value1",
}, {
    "Key": "ocim/instances/123/key2",
    "Value": "value2"
},{
    "Key": "ocim/instances/1234",
    "Value": "{\"key1\": \"value1\", \"key2\": \"value2\"}"
}
]```
And you run `consul.kv.get('ocim/instances/123')` then you will get an empty response because the exact key `ocim/instances/123` do not exists. If you run `consul.kv.get('ocim/instances/1234')` then the result will be:
```json
{
    "Key": "ocim/instances/1234",
    "Value": "{\"key1\": \"value1\", \"key2\": \"value2\"}"
}```
And if you do `consul.kv.get('ocim/instances/123', recurse=True)` then the response will be:
```json
[{
    "Key": "ocim/instances/123/key1",
    "Value": "value1",
}, {
    "Key": "ocim/instances/123/key2",
    "Value": "value2"
}]```

This is because consul's KV is built around the idea of prefixes in keys (when using `recurse=True`).

Reply:
> when using recurse=True

Oh, somehow I had in my mind both were fetched using that :man_facepalming: 

All's good.